File: task/revlearn.m
function revlearn(outfile,flag,varargin)
% code to perform reversal learning test
% saves data to outfile
% flag = 0 for generating a new set of switches; if so, varargin has two
% arguments, the number of contingency switches and the minimum and maximum 
% numbers of trials between switches
% flag = 1 loads data from a file (name given in varargin)

task = 'revlearn';

% default values if flag not specified
if ~exist('flag','var')
    nswitch = 4;
    minrun = 3;
    maxrun = 6;
    seed = 12345;
    trialvec = makeswitches(nswitch,minrun,maxrun,seed);
else
    % set up switches when flag is specified
    switch flag
        case 0
            nswitch=input('How many switches would you like?', 's');
            minrun=input('Minimum run?', 's');
            maxrun=input('Maximum run?', 's');
            seed = GetSecs;
            trialvec = makeswitches(nswitch,minrun,maxrun,seed);
        case 1
            dat = load(varargin{1});
            trialvec = dat.(varargin{2});
    end
    
end

% Unify Key Names
KbName('UnifyKeyNames'); %keynames will match those on Mac OS-X operating sys
Rkey=KbName('RightArrow');
Lkey=KbName('LeftArrow');
livekeys = [Lkey, Rkey];

%%%%%%%% PTB preliminaries %%%%%%%%%%%%%
PTBprelims

%%%%%%%%%%%%%% Sound Parameters %%%%%%%%%%%%%
setup_audio
[wrongsnd,buzzF]=wavread('buzz1.wav');
[rightsnd,dingF]=wavread('ding3.wav');

%%%%%%%%%%%%%% Task Parameters %%%%%%%%%%%%%
iti_mean = 3;
iti_range = 2;
pars.iti_mean = iti_mean;
pars.iti_range = iti_range;

%%%%%%%%%%%%%% Display Instructions %%%%%%%%%%%%%
instructions = {['Press the left or right key \n' ...
    'when the cross appears onscreen.\n' ...
    'You must learn by trial and error\n' ...
    'which is correct. \n\n' ...
    '(Press any key to continue)']};
display_instructions(win, instructions);

% Sound samples
txt = {'You will hear this for correct responses. \n (Press any key to continue)'};
playsound(pahandle, rightsnd);
display_instructions(win,txt);
txt = {'And this for incorrect responses. \n (Press any key to continue)'};
playsound(pahandle, wrongsnd);
display_instructions(win,txt);
txt = {'Please make a choice as soon as the "+" appears'; 'Press any key when ready.'};
display_instructions(win,txt);

%display onscreen countdown
countdown
ListenChar(2);

%%%%%%%% Start the Task %%%%%%%%%%%%%%%%%%%%%
tetio_startTracking;

WaitSecs(2); %give subjects a delay before first trial

for ind = 1:length(trialvec)
    %cue onset
    fixcross
    
    cuetime = tetio_localToRemoteTime(tetio_localTimeNow());
    
    % Wait for response
    [choice, ~] = handle_input(livekeys);
    presstime=tetio_localToRemoteTime(tetio_localTimeNow());
    
    % correct or incorrect?
    if (choice==Lkey && trialvec(ind) == 0) || (choice==Rkey && trialvec(ind) == 1)
        soundtime=tetio_localToRemoteTime(tetio_localTimeNow());
        playsound(pahandle,rightsnd);
        correct = 1;
    else
        soundtime=tetio_localToRemoteTime(tetio_localTimeNow());
        playsound(pahandle,wrongsnd);
        correct = 0;
    end
    
    WaitSecs(1); %outcome period
    
    % blank screen 
    Screen('FillRect',win, [0 0 0]);
    Screen('Flip',win);
    
    iti = iti_mean + iti_range*(2*rand-1);
    
    WaitSecs(iti);
    
    %%%% save data each trial %%%%%%
    data(ind).choice = choice;
    data(ind).correct = correct;
    data(ind).cuetime = cuetime;
    data(ind).presstime = presstime;
    data(ind).soundtime = soundtime;
    save(outfile,'data','task','pars')
     
end
tetio_stopTracking;

%read eye data
[lefteye, righteye, timestamp, trigSignal] = tetio_readGazeData;
eyedata.lefteye = lefteye;
eyedata.righteye = righteye;
eyedata.timestamp = timestamp;
eyedata.trig = trigSignal;

save(outfile,'data','eyedata','task','pars','trialvec')

shutdown_audio;
ListenChar(0);

% return the screen to dark
Screen('FillRect',win,[0 0 0],[]);
Screen('Flip',win);

disp('Program finished.');
end

File: pytask/revlearn.py
import numpy as np
from psychopy import prefs
prefs.general['audioLib'] = ['pyo']
from psychopy import visual, core, sound, event, gui
import display
import os


def revSettings(controller):
    vector_path = os.path.join(controller.settings_path, 'RevVectors.csv')
    # checks if previous vector file exists
    if os.path.isfile(vector_path):
        generateDlg = gui.Dlg(title="Reversal Learning")
        generateDlg.addField('Load previous (0) or Generate (1) Vectors', 0)
        generateDlg.show()
        if generateDlg.OK:
            response = generateDlg.data[0]
            if response == 0:
                return np.genfromtxt(vector_path, delimiter=',')
        else:
            return [-999]

    settingsDlg = gui.Dlg(title="Reversal Learning")
    settingsDlg.addText('Set Parameters')
    settingsDlg.addField(
        'Number of Switches', controller.settings['RevLearn: Number of Switches'])
    settingsDlg.addField(
        'Minimum Between', controller.settings['RevLearn: Minimum Between'])
    settingsDlg.addField(
        'Maximum Between', controller.settings['RevLearn: Maximum Between'])
    settingsDlg.show()  # show dialog and wait for OK or Cancel
    if settingsDlg.OK:
        response = settingsDlg.data
        return makeswitches(
            vector_path, response[0], response[1], response[2], np.random.randint(1, 9999))
    else:
        return [-999]


def revlearn(controller, outfile):
    trialvec = revSettings(controller)
    if trialvec[0] == -999:
        return
    # set up window
    # Create window to display test
    testWin = controller.launchWindow()

    # load sounds
    resource_path = '../task/'
    wrongsnd = sound.Sound(resource_path + 'buzz1.wav')
    rightsnd = sound.Sound(resource_path + 'dinga.wav')

    # parameters for task
    iti_mean = 3
    iti_range = 2

    # display instructions
    display.text_keypress(testWin,  ('Press the left or right key \n' +
                                     'when the cross appears onscreen.\n' +
                                     'You must learn by trial and error\n' +
                                     'which is correct. \n\n' +
                                     '(Press any key to continue)'))

    # sound samples
    rightsnd.play()
    display.text_keypress(
        testWin,  'You will hear this for correct responses. \n (Press any key to continue)')
    wrongsnd.play()
    display.text_keypress(
        testWin,  'And this for incorrect responses. \n (Press any key to continue)')
    display.text_keypress(
        testWin,  'Please make a choice as soon as the "+" appears\n(Press any key when ready)')

    # display countdown
    display.countdown(controller)

    # start eye tracking
    if not controller.testing:
        controller.tobii_cont.setDataFile(outfile)
        controller.tobii_cont.startTracking()
        controller.tobii_cont.setEventsAndParams(
            ['task', 'soundtime', 'presstime', 'cuetime', 'correct', 'choice', 'iti_mean', 'iti_range', 'trialvec'])
        controller.tobii_cont.setParam('task', 'revlearn')
        controller.tobii_cont.setParam('iti_mean', iti_mean)
        controller.tobii_cont.setParam('iti_range', iti_range)
        controller.tobii_cont.setVector('trialvec', trialvec)
    core.wait(2)

    for isTrue in trialvec:
        # display cross
        display.cross(testWin)
        if not controller.testing:
            controller.tobii_cont.recordEvent('cuetime')

        keypress = event.waitKeys(keyList=['left', 'right', 'q'])
        if not controller.testing:
            controller.tobii_cont.recordEvent('presstime')
            controller.tobii_cont.addParam('choice', keypress[0])
        if keypress[0] == 'q':
            break
        elif (keypress[0] == 'left' and not isTrue) or (keypress[0] == 'right' and isTrue):
            if not controller.testing:
                controller.tobii_cont.recordEvent('soundtime')
            rightsnd.play()
            correct = 1
        else:
            if not controller.testing:
                controller.tobii_cont.recordEvent('soundtime')
            wrongsnd.play()
            correct = 0
        if not controller.testing:
            controller.tobii_cont.addParam('correct', correct)
        # outcome period
        core.wait(1.0)
        # clear screen
        testWin.flip(clearBuffer=True)

        iti = iti_mean + iti_range * (2 * np.random.random() - 1)
        core.wait(iti)

    # stop eye tracking and save data
    if not controller.testing:
        controller.tobii_cont.stopTracking()
        controller.tobii_cont.flushData()
    testWin.close()


def makeswitches(path, nswitch, minrun, maxrun, seed):
    # set the seed
    np.random.seed(seed)

    # run lengths (between min and max) for each switch
    lens = np.random.random_integers(minrun, maxrun, nswitch + 1)
    # calculate switches
    switchtrials = np.cumsum(lens)
    isswitch = np.zeros((1, switchtrials[-1]))[0]
    isswitch[switchtrials[:-1]] = 1

    vals = np.zeros((1, nswitch + 1))[0]
    vals[1::2] = 1  # decides value at each switch

    # assign switch lengths to values
    whichval = np.cumsum(isswitch)
    trials = np.zeros((1, switchtrials[-1]))[0]
    for i in range(len(trials)):
        trials[i] = vals[whichval[i]]

    np.savetxt(path, trials, delimiter=",")
    return trials

if __name__ == '__main__':
    # revlearn(None,0,5,3,7)
    revlearn(None, 1)
